mypy/metastore.py

```diff
@@ -0,0 +1,172 @@
+"""Interfaces for accessing metadata.
+
+We provide two implementations.
+ * The "classic" file system implementation, which uses a directory
+   structure of files.
+ * A hokey sqlite backed implementation, which basically simulates
+   the file system in an effort to work around poor file system performance
+   on OS X.
+"""
+
+import binascii
+import sqlite3
+import os
+import time
+
+from abc import abstractmethod
+from typing import Dict, List, Set, Iterable, Any, Optional
+
+
+class MetadataStore:
+    @abstractmethod
+    def getmtime(self, name: str) -> float: ...
+
+    @abstractmethod
+    def read(self, name: str) -> str: ...
+
+    @abstractmethod
+    def write(self, name: str, data: str, mtime: Optional[float] = None) -> bool: ...
+
+    @abstractmethod
+    def remove(self, name: str) -> None: ...
+
+    @abstractmethod
+    def commit(self) -> None:
+        """If the backing store requires a commit, do it.
+
+        But N.B. that this is not *guarenteed* to do anything, just to be necessary
+        with the sqlite backend.
+        """
+        pass
+
+    @abstractmethod
+    def list_all(self) -> Iterable[str]: ...
+
+
+def random_string() -> str:
+    return binascii.hexlify(os.urandom(8)).decode('ascii')
+
+
+class FilesystemMetadataStore(MetadataStore):
+    def __init__(self, cache_dir_prefix: str) -> None:
+        self.cache_dir_prefix = cache_dir_prefix
```

Comment:
Currently the behavior is that the caches fail (via exceptions for read and an error code for write) when the cache dir is `/dev/null`. There's an explicit check for `SqliteMetadataStore` because otherwise the errors would be reported at cache creation time. The filesystem version reports the errors naturally by virtue of the `os` raising errors.

Is your suggestion that the caches should swallow errors when the cache is `/dev/null`, so that it doesn't require a check in `build`?

Reply:
I think it's fine that they fail, I just think it would be slightly tidier if they failed explicitly than by virtue of how /dev/null is a file not a directory or other arcana.